library/java/org/chromium/net/impl/CronetBidirectionalState.java

```diff
@@ -0,0 +1,454 @@
+    int CANCELLING = 1 << 8;            // EM's "cancel" method has been invoked.
+    int USER_CANCELLED = 1 << 9;        // The cancel operation was initiated by the User.
+    int FAILED = 1 << 10;               // An fatal failure has been encountered.
+    int ON_HEADER_RECEIVED = 1 << 11;   // EM's "onHeaders" callback has been invoked.
+    int ON_COMPLETE_RECEIVED = 1 << 12; // EM's "onComplete" callback has been invoked.
+    int READ_DONE = 1 << 13;            // User won't receive more read callbacks.
+    int WRITE_DONE = 1 << 14;           // User won't receive more write callbacks. Maybe never had.
+    int DONE = 1 << 15;                 // Terminal state. Can be successful or otherwise.
+    int TERMINATING_STATES = CANCELLING | FAILED | DONE; // Hold your breath and count to ten.
+  }
+
+  private final AtomicInteger mState = new AtomicInteger(State.NOT_STARTED);
+
+  /**
+   * Returns true if the final state has been reached. At this point the EM Stream has been
+   * destroyed.
+   */
+  boolean isDone() { return (mState.get() & State.DONE) != 0; }
+
+  /**
+   * Returns true if a terminating state has been reached. Terminating does not necessarily means
+   * that the DONE state has been reached. When the DONE bit is not set, it means that we are not
+   * ready yet to inform the user about the failure, as the EM as not yet destroyed the Stream. In
+   * other words, EM has not yet invoked a terminal callback (onError, onCancel, onComplete).
+   */
+  boolean isTerminating() { return (mState.get() & State.TERMINATING_STATES) != 0; }
+
+  /**
+   * Returns the reason why the request finished. Can only be invoked if {@link #isDone} returns
+   * true.
+   *
+   * @return one of {@link RequestFinishedInfo#SUCCEEDED}, {@link RequestFinishedInfo#FAILED}, or
+   *     {@link RequestFinishedInfo#CANCELED}
+   */
+  @FinishedReason
+  int getFinishedReason() {
+    assert isDone();
+    @State int finalState = mState.get();
+    if ((finalState & State.FAILED) != 0) {
+      return RequestFinishedInfo.FAILED;
+    }
+    if ((finalState & State.USER_CANCELLED) != 0) {
+      return RequestFinishedInfo.CANCELED;
+    }
+    return RequestFinishedInfo.SUCCEEDED;
+  }
+
+  /**
+   * Establishes what is the next action by taking in account the current global state, and the
+   * provided {@link Event}. This method has one important side effect: the resulting global state
+   * is saved through an Atomic operation. For few cases, this method will throw when the state is
+   * not compatible with the event.
+   */
+  @NextAction
+  int nextAction(@Event final int event) {
+    while (true) {
+      @State final int originalState = mState.get();
+
+      if (isAlreadyFinalState(event, originalState)) {
+        return NextAction.TAKE_NO_MORE_ACTIONS; // No need to loop - this is irreversible.
+      }
+
+      @NextAction final int nextAction; // "final" guarantees that it is assigned exactly once.
```

Comment:
This comment simply describes the java definition of final, right? I don't think that deserves a comment.

Reply:
Yeah - lately Java coders tend to remove the "final" because it is not rally needed. In this case, it is important to keep it for maintenance purposes.

Removed.